File: questions/78.Subsets.go
package questions

func subsets(nums []int) [][]int {
	var traverse func(i int)
	var result [][]int
	var subset []int

	traverse = func(i int) {
		if i >= len(nums) {
			result = append(result, append([]int{}, subset...))
			return
		}

		subset = append(subset, nums[i])
		traverse(i + 1)

		subset = subset[:len(subset)-1]
		traverse(i + 1)
	}

	traverse(0)

	return result
}

File: leetcode/78.Subsets.ts
function subsets(nums: number[]): number[][] {
  const result: number[][] = []
  const dfs = (startIndex: number, path: number[]) => {
    if (startIndex >= nums.length) {
      result.push([...path])
      return
    }

    path.push(nums[startIndex])
    dfs(startIndex + 1, path)

    path.pop()
    dfs(startIndex + 1, path)
  }

  dfs(0, [])

  return result
}
console.log(subsets([1, 2, 3])) // [ [ 1, 2, 3 ], [ 1, 2 ], [ 1, 3 ], [ 1 ], [ 2, 3 ], [ 2 ], [ 3 ], [] ]
